contracts/samples/VerifyingPaymaster.sol

```diff
@@ -52,26 +59,48 @@ contract VerifyingPaymaster is BasePaymaster {
     /**
      * verify our external signer signed this request.
      * the "paymasterAndData" is expected to be the paymaster and a signature over the entire request params
+     * paymasterAndData[:20] : address(this)
+     * paymasterAndData[20:28] : validUntil
+     * paymasterAndData[28:36] : validAfter
+     * paymasterAndData[36:] : signature
      */
     function validatePaymasterUserOp(UserOperation calldata userOp, bytes32 /*userOpHash*/, uint256 requiredPreFund)
-    external view override returns (bytes memory context, uint256 sigTimeRange) {
+    external override returns (bytes memory context, uint256 sigTimeRange) {
         (requiredPreFund);
 
         bytes32 hash = getHash(userOp);
+        uint256 nonce = senderNonce[userOp.getSender()]++;
         bytes calldata paymasterAndData = userOp.paymasterAndData;
-        uint256 sigLength = paymasterAndData.length - 20;
+        (uint64 validUntil, uint64 validAfter, bytes calldata signature) = parsePaymasterAndData(paymasterAndData);
+        uint256 sigLength = signature.length;
+
         //ECDSA library supports both 64 and 65-byte long signatures.
         // we only "require" it here so that the revert reason on invalid signature will be of "VerifyingPaymaster", and not "ECDSA"
         require(sigLength == 64 || sigLength == 65, "VerifyingPaymaster: invalid signature length in paymasterAndData");
 
         //don't revert on signature failure: return SIG_VALIDATION_FAILED
-        if (verifyingSigner != hash.toEthSignedMessageHash().recover(paymasterAndData[20 :])) {
-            return ("",1);
+        bytes32 digest = _hashTypedDataV4(keccak256(abi.encode(
+            keccak256("ValidatePaymasterUserOp(bytes32 hash,uint64 validUntil,uint64 validAfter,uint256 senderNonce)"),
```

Comment:
its a nice touch, to add support for the new validUntil/ValidFrom fields in the verification signature.
but why do you think we need EIP712  here? it is an internal signature, created by a backend service specifically for this contract.
I actually think its better to pass validFrom/validTo to getHash() method, to simplify the off-chain creation of the hash.

Reply:
I used EIP 712 for domain separator support. seems getHash() does not consider chainId and paymaster address which is weak point for signature replay.
But if we add chainid, paymaster address along with validUntil and validAfter on getHash() function, it will be enough.
I agree with simplification for off-chain so I can change this part to not use EIP712